app/controllers/marvin_js_assets_controller.rb

```diff
@@ -1,82 +1,89 @@
-      render json: new_asset
+        html: render_to_string(
+          partial: 'my_modules/result.html.erb',
+            locals: { result: result[:object] }
+        )
+      }, status: :ok
+    elsif result[:asset]
+      render json: result[:asset]
     else
-      render json: new_asset.errors, status: :unprocessable_entity
+      render json: result[:asset].errors, status: :unprocessable_entity
     end
   end
 
-  def show
-    sketch = current_team.marvin_js_assets.find_by_id(params[:id])
-    if sketch
-      if sketch.object_type == 'Step'
-        editable = can_manage_protocol_in_module?(sketch.object.protocol) ||
-                   can_manage_protocol_in_repository?(sketch.object.protocol)
-        render json: {
-          sketch: sketch,
-          editable: editable
-        }
-      else
-        render json: sketch
-      end
+  def update
+    asset = MarvinJsService.update_sketch(marvin_params, current_user)
+    if asset
+      render json: { url: rails_representation_url(asset.medium_preview), id: asset.id, file_name: asset.file_name }
     else
       render json: { error: t('marvinjs.no_sketches_found') }, status: :unprocessable_entity
     end
   end
 
-  def destroy
-    sketch = current_team.marvin_js_assets.find_by_id(params[:id])
-    if sketch.destroy
-      render json: sketch
+  private
+
+  def load_vars
+    @asset = current_team.assets.find_by_id(params[:id])
+    if action_name == 'create'
+      @assoc ||= Step.find_by_id(marvin_params[:object_id]) if marvin_params[:object_type] == 'Step'
+      @assoc ||= MyModule.find_by_id(params[:object_id]) if marvin_params[:object_type] == 'Result'
     else
-      render json: { error: t('marvinjs.no_sketches_found') }, status: :unprocessable_entity
+      return render_404 unless @asset
+
+      @assoc ||= @asset.step
+      @assoc ||= @asset.result
     end
-  end
 
-  def update
-    sketch = MarvinJsAsset.update_sketch(marvin_params, current_team)
-    if sketch
-      render json: sketch
-    else
-      render json: { error: t('marvinjs.no_sketches_found') }, status: :unprocessable_entity
+    if @assoc.class == Step
+      @protocol = @assoc.protocol
+    elsif @assoc.class == MyModule
+      @my_module = @assoc
+    elsif @assoc.class == Result
+      @my_module = @assoc.my_module
     end
   end
 
-  def team_sketches
-    result = ''
-    sketches = current_team.marvin_js_assets.where.not(object_type: 'TinyMceAsset')
-    sketches.each do |sketch|
-      result += render_to_string(
-        partial: 'shared/marvinjs_modal_sketch.html.erb',
-           locals: { sketch: sketch }
-      )
+  def check_read_permission
+    if @assoc.class == Step
+      render_403 && return unless can_read_protocol_in_module?(@protocol) ||
+                                  can_read_protocol_in_repository?(@protocol)
+    elsif @assoc.class == Result || @assoc.class == MyModule
+      render_403 and return unless can_read_experiment?(@my_module.experiment)
```

Comment:
Use && instead and, as you did in above case. Or you can simply do 'return render_403'